app/src/main/java/org/oppia/android/app/player/state/itemviewmodel/TextInputViewModel.kt

```diff
@@ -28,20 +30,46 @@ class TextInputViewModel private constructor(
 ) : StateItemViewModel(ViewType.TEXT_INPUT_INTERACTION), InteractionAnswerHandler {
   var answerText: CharSequence = ""
   val hintText: CharSequence = deriveHintText(interaction)
+  private var pendingAnswerError: String? = null
 
   var isAnswerAvailable = ObservableField<Boolean>(false)
+  val errorMessage = ObservableField<String>("")
 
   init {
     val callback: Observable.OnPropertyChangedCallback =
       object : Observable.OnPropertyChangedCallback() {
         override fun onPropertyChanged(sender: Observable, propertyId: Int) {
           interactionAnswerErrorOrAvailabilityCheckReceiver.onPendingAnswerErrorOrAvailabilityCheck(
-            /* pendingAnswerError= */ null,
-            answerText.isNotEmpty()
+            pendingAnswerError = pendingAnswerError,
+            inputAnswerAvailable = true // Allow submit on empty answer.
           )
         }
       }
     isAnswerAvailable.addOnPropertyChangedCallback(callback)
+    errorMessage.addOnPropertyChangedCallback(callback)
+
+    // Initializing with default values so that submit button is enabled by default.
+    interactionAnswerErrorOrAvailabilityCheckReceiver.onPendingAnswerErrorOrAvailabilityCheck(
+      pendingAnswerError = null,
+      inputAnswerAvailable = true
+    )
+  }
+
+  override fun checkPendingAnswerError(category: AnswerErrorCategory): String? {
+    when (category) {
+      AnswerErrorCategory.REAL_TIME -> {
+        pendingAnswerError = null
+      }
+      AnswerErrorCategory.SUBMIT_TIME -> {
+        pendingAnswerError =
+          TextParsingUiError.createFromParsingError(
+            answerText.toString()
+          ).getErrorMessageFromStringRes(resourceHandler)
+      }
+    }
+
+    errorMessage.set(pendingAnswerError)
+    return pendingAnswerError
```

Comment:
```suggestion
    return when (category) {
      AnswerErrorCategory.REAL_TIME -> null
      AnswerErrorCategory.SUBMIT_TIME -> {
        TextParsingUiError.createFromParsingError(
          answerText.toString()
        ).getErrorMessageFromStringRes(resourceHandler)
      }
    }.also {
      pendingAnswerError = it
      errorMessage.set(it)
    }
```

Slight alternative approach that uses some Kotlin idioms & forces branch checking if new categories are added in the future.

Reply:
Okay.